File: 01_simple_examples/python/04_cv+_libraries.py
#################################################
# CV + - Regression
# Libraries
#################################################

import time
import numpy as np
from sklearn.ensemble import RandomForestRegressor
from mapie.regression import CrossConformalRegressor

# Data Generation 
np.random.seed(1)
n = 100

x = np.random.uniform(0, 5, n)
y = x**2 + np.random.normal(0, 1.5, n)

# New Test Instance
x_new = 2.5
y_true = x_new**2 + np.random.normal(0, 1.5)

alpha = 0.10

#################################################
# MAPIE 
#################################################

start_time = time.time()

# Base model specification
model = RandomForestRegressor(random_state=0)

# Configure CV+ 
mapie = CrossConformalRegressor(
    estimator=model,
    method="plus",        
    cv=10,                
    confidence_level=1 - alpha   
)

# Model training and cross-residual computation
mapie.fit_conformalize(x.reshape(-1, 1), y)

end_time = time.time()

execution_time = end_time - start_time
print(f"MAPIE execution time: {execution_time:.4f} seconds")

# Prediction with intervals for the new test instance
y_pred, y_pis = mapie.predict_interval(np.array([[x_new]]))

# Prediction interval
print("MAPIE Library Interval:")
print(y_pis)

File: 01_simple_examples/R/04_cv+_libraries.R
#################################################
# CV+ - Regression
# Libraries
#################################################

# Data Generation
set.seed(1)
n <- 100  
x <- runif(n, 0, 5)
y <- x^2 + rnorm(n, sd = 1.5)

# New Test Instance 
x_new <- 2.5
y_true <- x_new^2 + rnorm(1, sd = 1.5)

alpha <- 0.10


#################################################
# Predictset
#################################################

library(predictset)

start_time <- Sys.time()

# Model and Wrapper Specification for Predictset 
rf_model <- make_model(
    train_fun = function(x, y) {
        set.seed(1)
        randomForest(x = data.frame(x = x), y = y)
    },
    predict_fun = function(object, x_new) {
        predict(object, newdata = data.frame(x = x_new))
    },
    type = "regression"
)

# Run Jackknife+
result <- conformal_cv(
    x = data.frame(x = x),
    y = y,
    model = rf_model,
    x_new = data.frame(x = x_new),
    alpha = 0.1,
    n_folds = 10
)

end_time <- Sys.time()

execution_time_ps <- as.numeric(end_time - start_time, units = "secs")
print(paste("Predictset execution time:", execution_time_ps, "seconds"))

prediction_interval_ps <- c(result$lower, result$upper)
print("Predictset Library Interval:")
print(prediction_interval_ps)